environment-views: fix bug where missing recipe/repo breaks env commands

robo-wylder opened the pull request:
Issue:
When a recipe or a repo has been removed from Spack and an environment
is active, it causes the view activation to crash Spack before any
commands can be executed. Further, the resulting error message is not clear
in explaining the issue or a workaround.

Solution:
Catch the error for both unknown packages and repos and emit a clear error
message with options of what to do.

Added a `--no-env-view` flag to the main entry point to skip view activation
for cases like this. Additionally, this forces view regeneration to always start
from scratch to ensure old links are cleaned up properly.

robo-wylder committed:
environment-views: fix bug where missing recipe/repo breaks env commands

When a recipe or a repo has been removed from Spack and an environment
is active, it causes the view activation to crash Spack before any
commands can be executed. Further, the error message it not at all clear
in explaining the issue.

This forces view regeneration to always start from scratch to avoid the
missing package recipes, and also provides a CLI option to not activate
the view during a command.

becker33 commented on lib/spack/spack/main.py:
```diff
@@ -709,7 +712,7 @@ def main(argv=None):
         env = ev.find_environment(args)
         if env:
-            ev.activate(env, args.use_env_repo)
+            ev.activate(env, args.use_env_repo, not args.no_env_view)
```
I don't think we need an argument here, this should never activate the view with the environment. Commands other than `spack env activate` and `spack env deactivate` aren't equipped to handle the relevant environment changes, and those commands activate/deactivate the environment later (in the SpackCommand()()) call.

robo-wylder replied:
If we always pass `add_view=False`, then detecting this issue works well when trying to activate an environment with a missing package/namespace, but it won't detect when this happens and you're already in the environment. Maybe we should have the `spack env status` command also validate that all packages still exist?

becker33 commented on lib/spack/spack/environment.py:
```diff
@@ -175,9 +175,17 @@ def activate(
+            spack.repo.UnknownNamespaceError) as e:
+        tty.error(e)
+        tty.die(
+            'Environment DAG is broken due to a missing package or repo.',
+            'To resolve without it, force concretize with the command: ',
+            '   spack --no-env-view concretize --force')
```
```suggestion
            'Environment view is broken due to a missing package or repo.\n',
            '  To remove it and resolve the issue, ',
            'force concretize with the command:\n',
            '    spack concretize --force')
```

becker33 replied:
We use DAG to refer to the dependency graph among packages, I think calling it the `Environment view` will be less confusing to users here.

robo-wylder replied:
Ok, that makes sense. I think technically the DAG would be broken too though because the package it concretized with no longer exists.

becker33 replied:
We can't recreate it, but as long as we don't need to look up package information (like recalculating the run environment) we can still use that DAG. For example, the user could use `spack env loads` to generate a script that loads all the modules, and the DAGs will still be valid when used in that way.

becker33 reviewed (dismissed):
I think we can just go a step further and just always add a `False` argument instead of having an option. That allows us to make a cleaner suggestion to the user as well.

robo-wylder committed:
add messages to env status and deactivate

Remove the --no-env-view option and default add_view=False in main.

Warn users that a view may be corrupt when deactivating an environment
or checking its status while active. Updated message for activate.

becker33 commented:
@robo-wylder can you remove the changes to the bash completion script? That's what's holding up the tests.

robo-wylder committed:
remove argument from bash completion script

robo-wylder commented:
Updated the completion script, though I had to do so by hand as the `spack commands --update-completion` did not update that section. I think it has something to do with the command being in the `env-group`

robo-wylder committed:
adjust string format call to be compatible with 2.6

becker33 commented:
Closing and reopening to restart tests

becker33 reviewed (approved):
Might end up needing a unit test. If you want I can try to add that

robo-wylder commented:
> Might end up needing a unit test. If you want I can try to add that

I'll try to get that in today or tomorrow but if I can't I'm on vacation next week so you may have to.

becker33 committed:
tests for view checking

becker33 commented:
@robo-wylder I pushed some tests this morning; I'm trying to cut the release around noon.
